zebra/zebra_mpls.c

```diff
@@ -2999,11 +3003,30 @@ int zebra_mpls_write_label_block_config(struct vty *vty, struct zebra_vrf *zvrf)
 /*
  * Called when VRF becomes inactive, cleans up information but keeps
  * the table itself.
- * NOTE: Currently supported only for default VRF.
  */
 void zebra_mpls_cleanup_tables(struct zebra_vrf *zvrf)
 {
-	hash_iterate(zvrf->lsp_table, lsp_uninstall_from_kernel, NULL);
+	if (zvrf_id(zvrf) == VRF_DEFAULT)
+		hash_iterate(zvrf->lsp_table, lsp_uninstall_from_kernel, NULL);
+	else {
+		/*
+		 * For other vrfs, we try to remove associated LSPs; we locate
+		 * the LSPs in the default vrf.
+		 */
+		struct zebra_vrf *def_zvrf;
+		afi_t id;
```

Comment:
nit: I think the `id` name is a bit unusual here, I'd prefer a simple `afi` to be consistent with the rest of the code base. Also, this variable can be declared in the control part of the `for` loop below in order to reduce its scope.

Reply:
got it - I've pushed a change to move and rename the locals.

agree about the zapi client issue - will look at that too.

Reply:
Looks like a hook function is registered for zapi clients, so I think that case is ok. see zebra_mpls.c:zebra_mpls_cleanup_fecs_for_client()